### ISSUE TITLE
bcrypt / hash function evolution

### ISSUE THREAD

**Issue by Indigo744** [benedmunds/CodeIgniter-Ion-Auth#1195]
Dear @benedmunds 

First and foremost, thank you again for this lib. I use it extensively and I love it  :smile: This is why I try to give back by proposing updates and enhancement when I find the time.

The usage of the specific bcrypt library has always bugged me. Of course I've audited the code and found out that it was OK using `crypt()` and such, but I keep thinking that it could be better.

I found out that I was not the only one (see related issues #1176 and #1169). 

Another thing is using `crypt()` function will not allow us to provide an option to use the new ARGON2 algorithm, which is available in PHP 7.2 but only for the `password_hash()` function.

Hence I would like to make a PR to move to using `password_*` functions.

But before starting to work on it, I have a couple of questions I would like to pass by you first:
1. Should PHP < 5.5 still be supported? [According to Wikipedia](https://en.wikipedia.org/wiki/PHP#Release_history), PHP 5.3 is not supported since 2014, and PHP 5.5 since 2016... 
Should an Auth library really support this old version?
2. If PHP < 5.5 is needed, should we use [password_compat](https://github.com/ircmaxell/password_compat)? 
How should we load it? Copy/pasting the code would seem easy, but it would be troublesome to maintain (checking for updates regularly, etc...). 
Also, CI 3 offers their own polyfills (see `core/compat/password.php`) so it would be quite redundant...
So maybe we should simply redirect users to this lib and recommends them to install it? I really wonder the number of users with CI 2 and PHP < 5.5...

Thanks for your insights.

**Comment by Indigo744** [comment]
Also, there are two differents salt generator functions implemented: `Ion_auth_model::salt()` and `Bcrypt::getSalt()`...

**Comment by benedmunds** [comment]
Hey @Indigo744 

Pretty excited that you're up for taking this on.  

I'm thinking that this would be worth cutting a new major version of Ion Auth (v3) and then you could just support CI3 and rely on the polyfills being present.  

This versioning makes sense and will support adding a CI 4 version later as well.

What are your thoughts on the above?

**Comment by Indigo744** [comment]
So it would mean we should target CI 3 and PHP 5.5?
[CI 3 server requirement](https://www.codeigniter.com/user_guide/general/requirements.html) ask for PHP 5.6, but suggest 5.3.7 would work (but strongly advised against).

I think that would be a good thing indeed, and would allow to clean up some old legacy code.

How should we approach this? Maybe open a new branch so I can push a PR to it?
README will need a proper clean up as well 😅 

The `2` branch could be put in security maintenance mode (i.e. only implement security related issues).

**Comment by Indigo744** [comment]
As another matter, I really think the md5/sha1 fallback should be dropped. They are clearly proved insecure and should never be used for password storage. 

I really think Ion Auth, as an Auth library, should enforce good security standards.

What do you think?

**Comment by benedmunds** [comment]
I'm cool with PHP 5.6+

I created a "3" branch, you can PR against that: https://github.com/benedmunds/CodeIgniter-Ion-Auth/tree/3

Agreed on that, I won't merge anything new aside from security/bug fixes.

**Comment by Indigo744** [comment]
Awesome. I'll start working on it. I may send multiple PR depending on the scope, to allow proper review.

**Comment by benedmunds** [comment]
Sounds great, thank you!

**Comment by Indigo744** [comment]
By the way, what are your thoughts on [my comment about sha1 support](https://github.com/benedmunds/CodeIgniter-Ion-Auth/issues/1195#issuecomment-374360036)?

**Comment by benedmunds** [comment]
Oh sorry, missed that.  I'm fine with dropping support for it for new user registrations but I think it'd be a good idea to add migration instructions (or a method with controller implementation) for a password-needs-rehash type functionality to upgrade old users from SHA1 to BCrypt.

Not sure the best way to do that, maybe check that the hash matches BCrypt format and assume it needs rehash if it doesn't?  Then still validate the SHA pass but then upgrade it?

Hmm, that will add a lot of complexity though.  It might be worth just dropping with no migration and then writing a readme for it.  What do you think?

**Comment by Indigo744** [comment]
Indeed it could be good to offer migration capability.

I'll see if it isn't too hard. I'll let you know ;-)

**Comment by benedmunds** [comment]
Awesome 👏

**Comment by Indigo744** [comment]
I was looking at [your review](https://github.com/benedmunds/CodeIgniter-Ion-Auth/pull/1196#pullrequestreview-105152878) and it struck me that the term "rounds" is actually not really true. This option indicates the `cost` value, which leads to `2^cost` number of rounds.

Should we rename it (e.g. `default_rounds` to `bcrypt_cost`), in order to be align to the existing literature? And allow further option when we'll add more algorithm?

**Comment by benedmunds** [comment]
Yea good point, go for it

**Comment by Indigo744** [comment]
Ok, this is progressing nicely.

Next things on my todo list:

- [x]  implement Argon2 algo option (PHP 7.2)
- [x]  implement different algo params depending on admin or not
- [x]  implement password_need_rehash check (to have smooth transition between costs and algo)
- [x]  implement sha1 BC
- [x]  Improve "Remember me" and "forgotten pwd" security
- [x]  Make sure  "Remember me" and "forgotten pwd" fields are always clear on state change
- [x]  Update doc
- [x]  Update Migration script

**Comment by benedmunds** [comment]
Thoughts are switching to short array syntax in v3 since the min PHP version has been bumped?  

Probably worth seeing what else we can clean up syntax-wise.

**Comment by Indigo744** [comment]
The short array syntax was [introduced ](http://php.net/releases/5_4_0.php) in PHP 5.4.
Since it is not polyfill-able  (is that a word? 😄), it really depends on if you want to allow users to deploy Ion Auth 3 on PHP < 5.4. It's quite a stretch indeed so I think it's safe to use it.

I don't see anything else right now.

**Comment by benedmunds** [comment]
Yea I think I'm fine with forcing 5.4+

**Comment by Indigo744** [comment]
As for testing, when all PR will be merged, we should perform our own testing. But maybe we can also open a new issue to call for help? 😅

**Comment by benedmunds** [comment]
Yea definitely!

**Comment by Indigo744** [comment]
Side note: I've opened [a StackOverflow question](https://stackoverflow.com/questions/49403437/argon2-algorithm-in-php7-understanding-the-time-cost-parameter) regarding Argon2 and its (confusing) time_cost parameters.

**Comment by benedmunds** [comment]
Ok cool, I commented on the SO post.

**Comment by Indigo744** [comment]
Damn, now I feel silly to have asked on SO 😓 . I should have looked in the code. Well, it may help other people looking for an answer...

**Comment by benedmunds** [comment]
Oh don't feel silly at all.  It was a great question and something that will definitely help others later.

**Comment by Indigo744** [comment]
I'll wait a bit before selecting your answer. I just want to see if some people want to chime in with some valuable information.

**Comment by benedmunds** [comment]
Sounds good 👍

**Comment by Indigo744** [comment]
OK, I've checked and here is the evolution of functions in `Ion_auth_model.php`

## Functions removed  

```php
protected function _prepare_ip($ip_address) 
protected function _regenerate_session()

public function forgotten_password_complete($code, $salt = FALSE)
public function hash_code($password)
public function is_time_locked_out($identity, $ip_address = NULL)
public function salt()
```

## Functions added

```php
protected function _generate_selector_validator_couple($selector_size = 40, $validator_size = 128)
protected function _get_hash_algo()
protected function _get_hash_parameters($identity = NULL)
protected function _password_verify_sha1_legacy($identity, $password, $hashed_password_db)
protected function _random_token($result_length = 32)
protected function _retrieve_selector_validator_couple($user_code)
protected function _set_password_db($identity, $password)

public function db()
public function clear_remember_code($identity) 
public function get_user_by_forgotten_password_code($user_code)
public function get_user_id_from_identity($identity = '')
public function rehash_password_if_needed($hash, $identity, $password)
```

## Functions modified

```php
public function hash_password_db($id, $password, $use_sha1_override = FALSE)
/* TO */
public function verify_password($password, $hash_password_db, $identity = NULL)
```

```php
public function clear_forgotten_password_code($code)
/* TO */
public function clear_forgotten_password_code($identity)
```

```php
public function hash_password($password, $salt = FALSE, $use_sha1_override = FALSE)
/* TO */
public function hash_password($password, $identity = NULL)
```

```php
public function remember_user($id)
/* TO */
public function remember_user($identity)
```

**Comment by Indigo744** [comment]
Here are all functions using `$id`, and all functions using `$identity`.

# For Ion Auth 2 

## Functions using $id

```php
public function hash_password_db($id, $password, $use_sha1_override = FALSE)
public function activate($id, $code = FALSE)
public function deactivate($id = NULL)
public function user($id = NULL)
public function get_users_groups($id = FALSE)
public function add_to_group($group_ids, $user_id = FALSE)
public function remove_from_group($group_ids = FALSE, $user_id = FALSE)
public function group($id = NULL)
public function update($id, array $data)
public function delete_user($id)
public function update_last_login($id)
public function remember_user($id)
```

## Functions using $identity

```php
public function reset_password($identity, $new) 
public function change_password($identity, $old, $new)
public function identity_check($identity = '')
public function forgotten_password($identity)
public function register($identity, $password, $email, $additional_data = array(), $groups = array())
public function login($identity, $password, $remember=FALSE)
public function is_max_login_attempts_exceeded($identity, $ip_address = NULL)
public function get_attempts_num($identity, $ip_address = NULL)
public function is_time_locked_out($identity, $ip_address = NULL)
public function get_last_attempt_time($identity, $ip_address = NULL)
public function get_last_attempt_ip($identity)
public function increase_login_attempts($identity)
public function clear_login_attempts($identity, $old_attempts_expire_period = 86400, $ip_address = NULL)
```

# For Ion Auth 3 (now)

## Functions using $id
```php
public function activate($id, $code = FALSE)
public function add_to_group($group_ids, $user_id = FALSE)
public function deactivate($id = NULL)
public function delete_user($id)
public function get_users_groups($id = FALSE)
public function group($id = NULL)
public function in_group($check_group, $id = FALSE, $check_all = FALSE)
public function remove_from_group($group_ids = FALSE, $user_id = FALSE)
public function update($id, array $data)
public function update_last_login($id)
public function user($id = NULL)
```

## Functions using $identity
```php
protected function _get_hash_parameters($identity = NULL)
protected function _password_verify_sha1_legacy($identity, $password, $hashed_password_db)
protected function _set_password_db($identity, $password)
public function change_password($identity, $old, $new)
public function clear_forgotten_password_code($identity) 
public function clear_login_attempts($identity, $old_attempts_expire_period = 86400, $ip_address = NULL)
public function clear_remember_code($identity) 
public function forgotten_password($identity)
public function get_attempts_num($identity, $ip_address = NULL)
public function get_last_attempt_ip($identity)
public function get_last_attempt_time($identity, $ip_address = NULL)
public function get_user_id_from_identity($identity = '')
public function hash_password($password, $identity = NULL)
public function identity_check($identity = '')
public function increase_login_attempts($identity)
public function is_max_login_attempts_exceeded($identity, $ip_address = NULL)
public function login($identity, $password, $remember=FALSE)
public function register($identity, $password, $email, $additional_data = array(), $groups = array())
public function rehash_password_if_needed($hash, $identity, $password)
public function remember_user($identity)
public function reset_password($identity, $new) 
public function verify_password($password, $hash_password_db, $identity = NULL)
```

**Comment by Indigo744** [comment]
Ok I tried to gather as much as information as I could. What do you think?

**Comment by benedmunds** [comment]
Gotcha, this seems fine really... Most of these identity ones won't be used my user-land often.

**Comment by Indigo744** [comment]
I don't know. Actually I didn't really understand the use case for multiple identity? Do you have an example?

**Comment by benedmunds** [comment]
A lot of sites let people login with either a username or an email address. In the past I’ve seen people accomplish this with Ion Auth by dynamically changing the config identity to be whichever and trying login with each.

**Comment by Indigo744** [comment]
Oh ok. I wonder what would be the best approach to handle this natively without resorting to that kind of thing...

**Comment by benedmunds** [comment]
Maybe make identity an array? and change all the queries that use it to build orWheres... I think that's all that would be needed...

**Comment by Indigo744** [comment]
Sadly, this doesn't seems _that_ simple... The `identity` option is used to use the `$this->identity_column`, which is used quite a lot in different function... Which would need to be updated as well....

**Comment by benedmunds** [comment]
Yea good point...  I doubt it's worth it either looking at this.  We'd need to refactor all of the methods except login and register to use ID.  but then it'll cause other performance concerns.

**Comment by Indigo744** [comment]
If a user come asking for this, we could have another look, but for now this seems quite sufficient 😸

**Comment by Indigo744** [comment]
I think only the doc is remaining. You can start reviewing the PR if you have some time.

**Comment by Indigo744** [comment]
Side note: what do you think about a changelog.md?

**Comment by benedmunds** [comment]
Yea good idea, especially since there could be possible breaking changes here

**Comment by Indigo744** [comment]
I've already written quite some explanation in [UPGRADING.md](https://github.com/Indigo744/CodeIgniter-Ion-Auth/blob/update_doc/UPGRADING.md).
I think the changelog should just list the new/removed features?

**Comment by Indigo744** [comment]
Another question: `min_password_length` and `max_password_length` are not actually enforced by the library itself. Maybe we should add a note or something (like for `maximum_login_attempts`).

**Comment by benedmunds** [comment]
Yea, what do you think about removing max password length and then enforcing min length at the library level?

A max length doesn’t really make sense anymore, we should encourage people to use pass phrases and password tools like 1password.

**Comment by Indigo744** [comment]
Indeed. I hate when a website tells me that my Keepass-generated password is too long...!
Also, in my app with Ion Auth, I only enforce the min password at validation level (when validating form input).
However, enforcing min length at lib level would mean adding a new sentence in the lang files.
Also, I found out recently that, using bcrypt, only the first 72 characters are actually used for computing the hash (this seems to be due to its internal bloc size, see [here for more information](https://security.stackexchange.com/questions/39849/does-bcrypt-have-a-maximum-password-length)).
Some people actually suggests to use an "intermediate hash" (something like `bcrypt(sha*(pass))` to get the max entropy out of every password. From [this article](https://paragonie.com/blog/2016/02/how-safely-store-password-in-2016#why-scrypt), it is suggested `bcrypt(base64_encode(sha384(password)))`.
However, some other people suggest that doing so, would likely reduce entropy for large password (such as the one generated with a password manager). Small or trivial password, in the other hand, may benefit from a increase of entropy.

**Comment by benedmunds** [comment]
Ha yea it’s super annoying when that happens.

Yea I’m aware of the truncation, I’m on the side of its best to let bcrypt truncate and not screw with the password cause that does add additional risks.

**Comment by Indigo744** [comment]
I'm with you on this. And if people don't want bcrypt to truncate their passwords, they should use argon2, which doesn't have any limitations.

However, should we say something about it in the doc?

**Comment by benedmunds** [comment]
I’m not against saying something but also don’t think it’s neccesary, common practice (at the moment at least) is that it’s plenty secure even with the truncation.

**Comment by Indigo744** [comment]
I think we can see the end 😄  what do you think?

We should now try it out as much as possible.

**Comment by Indigo744** [comment]
Should we consider namespaces?

I know that CI4 will use them extensively. Maybe we can try to prepare for the future?
Actually, I'm not even sure it would work with how CI3 is designed.
This is just a thought.

**Comment by benedmunds** [comment]
I think we can skip them for v3.  We should go ahead and start working on a v4 version once this clears some testing though.  That's coming soon.

**Comment by Indigo744** [comment]
Alright. Should I open a new issue to call for some help and testing from users?

**Comment by benedmunds** [comment]
Yea go for it :D  I'll tweet it out and hopefully get some people involved

**Comment by Indigo744** [comment]
See #1220

**Comment by benedmunds** [comment]
Closing this ticket so we can consolidate feedback in #1220 

I believe this is about ready to be the default branch.  Waiting on some feedback in #1220.
